### PR TITLE
[Remote Store] Emit correct global checkpoint during translog upload

### PR DESCRIPTION
**Pull request by ashking94**
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ashking94** [comment]
> ❌ Gradle check result for [a1d5a87](https://github.com/opensearch-project/OpenSearch/commit/a1d5a8760272131b1a1502d59b36436ffa9ab477): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48058/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

I have added tests for the edge case that is mentioned in the referenced issue. I had added the tests first and then the main code changes -
```
[org.opensearch.remotestore.RemoteStoreIT.testCloseIndexWithNoOpSyncAndFlushForSyncTranslog](https://build.ci.opensearch.org/job/gradle-check/48058/testReport/junit/org.opensearch.remotestore/RemoteStoreIT/testCloseIndexWithNoOpSyncAndFlushForSyncTranslog/)
[org.opensearch.remotestore.RemoteStoreIT.testCloseIndexWithNoOpSyncAndFlushForSyncTranslog](https://build.ci.opensearch.org/job/gradle-check/48058/testReport/junit/org.opensearch.remotestore/RemoteStoreIT/testCloseIndexWithNoOpSyncAndFlushForSyncTranslog_2/)
[org.opensearch.remotestore.RemoteStoreIT.testCloseIndexWithNoOpSyncAndFlushForSyncTranslog](https://build.ci.opensearch.org/job/gradle-check/48058/testReport/junit/org.opensearch.remotestore/RemoteStoreIT/testCloseIndexWithNoOpSyncAndFlushForSyncTranslog_3/)
[org.opensearch.remotestore.RemoteStoreIT.testCloseIndexWithNoOpSyncAndFlushForSyncTranslog](https://build.ci.opensearch.org/job/gradle-check/48058/testReport/junit/org.opensearch.remotestore/RemoteStoreIT/testCloseIndexWithNoOpSyncAndFlushForSyncTranslog_4/)
```

**Comment by ashking94** [comment]
> ❌ Gradle check result for [29cf87f](https://github.com/opensearch-project/OpenSearch/commit/29cf87fcee64cef936b052b5f499acc373707b32): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48078/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Fixed this.

**Review by Bukhtawar — COMMENTED** [review]
Wondering is this the problem with just remote translog or local translog as well?

**Comment by ashking94** [comment]
> Wondering is this the problem with just remote translog or local translog as well?

The problem seems to exist for remote translog only since the local version seems fine, but when close the index, the translog is wiped out locally and rehydrated from remote store. At the point, the most recent checkpoint file downloaded has a global checkpoint from the last but 1 translog sync.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java [hunk 1]
```diff
@@ -1011,4 +1012,73 @@ public void testAsyncTranslogDurabilityRestrictionsThroughIdxTemplates() throws
...
+            indexSingleDoc(INDEX_NAME);
+            indexSingleDoc(INDEX_NAME);
+            // Reduce the latch for the main thread to flush after some sleep.
+            latch.countDown();
+        }).start();
+        // Wait for atleast one doc to be ingested.
+        latch.await();
+        // Sleep for some time for the next doc to be present in lucene buffer. If flush happens first before the doc #2
+        // gets indexed, then it goes into the happy case where the close index happens succefully.
+        Thread.sleep(1000);
```
Comment (gbbafna) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:1076]: curious why do we need sleep here ? As we have already indexed docs and they should be in lucene buffer.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java [hunk 2]
```diff
@@ -1011,4 +1012,73 @@ public void testAsyncTranslogDurabilityRestrictionsThroughIdxTemplates() throws
...
+        // Flush so that the subsequent sync or flushes are no-op.
+        flush(INDEX_NAME);
+        // Closing the index involves translog.sync and shard.flush which are now no-op.
+        client().admin().indices().close(Requests.closeIndexRequest(INDEX_NAME)).actionGet();
+        Thread.sleep(10000);
+        ensureGreen(INDEX_NAME);
+    }
+
+    public void testCloseIndexWithNoOpSyncAndFlushForAsyncTranslog() throws InterruptedException {
```
Comment (gbbafna) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:1055]: can we write UTs as well for same ?

**Review by gbbafna — COMMENTED** [review]
Changes looks great .

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java (hunk 1 above)
Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:1076]: This is not required, got left behind when copying the async test from other sync test that I have added. But having it does not have any side effects too. Removing the sleep.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java (hunk 2 above)
Comment (ashking94) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreIT.java:1055]: Added UTs as well.

**Review by gbbafna — APPROVED** [review]
